File: backend/app/schemas/application.py
import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from app.models.application import ApplicationStatus


class ApplicationBase(BaseModel):
    company: str = Field(min_length=1, max_length=255)
    title: str = Field(min_length=1, max_length=255)
    location: str | None = Field(default=None, max_length=255)
    job_url: str | None = None
    status: ApplicationStatus = ApplicationStatus.SAVED
    job_description: str = Field(min_length=1)
    notes: str | None = None


class ApplicationCreate(ApplicationBase):
    pass


class ApplicationUpdate(BaseModel):
    company: str | None = Field(default=None, min_length=1, max_length=255)
    title: str | None = Field(default=None, min_length=1, max_length=255)
    location: str | None = Field(default=None, max_length=255)
    job_url: str | None = None
    status: ApplicationStatus | None = None
    job_description: str | None = Field(default=None, min_length=1)
    notes: str | None = None


class ApplicationRead(ApplicationBase):
    id: uuid.UUID
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)

File: frontend/src/types/application.ts
export type ApplicationStatus =
  | "SAVED"
  | "APPLIED"
  | "INTERVIEW"
  | "OFFER"
  | "REJECTED";

export interface Application {
  id: string;
  company: string;
  title: string;
  location: string | null;
  job_url: string | null;
  status: ApplicationStatus;
  job_description: string;
  notes: string | null;
  created_at: string;
  updated_at: string;
}